### ISSUE TITLE
numeric overflow in InvalidatingNodeVisitor.DeletingNodeVisitor

### ISSUE THREAD

**Issue by neumannt** [bazelbuild/bazel#19445]
### Description of the bug:

The function `runInternal` in `InvalidatingNodeVisitor.DeletingNodeVisitor` contains the following fragment:

```
        int numThreads = min(DEFAULT_THREAD_COUNT, listSize);
        for (int i = 0; i < numThreads; i++) {
          int index = i;
          executor.execute(
              () ->
                  visit(
                      Collections2.transform(
                          pendingList.subList(
                              (index * listSize) / numThreads,
                              ((index + 1) * listSize) / numThreads),
                          Pair::getFirst),
                      InvalidationType.DELETED));
        }
```

In a very large project that is build on a machine with many cores the expression `((index + 1) * listSize) / numThreads)` sometimes becomes negative, presumably due to a numeric overflow. Which then crashes in the call to `subList`.

### Which category does this issue belong to?

Core

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

Unfortunately I do not have a reproducer. This happens non-deterministically, and only in builds that have already ran for 1 1/2 hours. But I have included a stack trace below, and it shows that the second argument to `subList` becomes negative.

### Which operating system are you running Bazel on?

Linux, version 3.10.0-1160.95.1.el7.x86_64

### What is the output of `bazel info release`?

release 6.3.2

### Any other information, logs, or outputs that you want to share?

The stack trace in the moment of the crash:

```
FATAL: bazel crashed due to an internal error. Printing stack trace:
    java.lang.IndexOutOfBoundsException: end index (-29676970) must not be negative
    at com.google.common.base.Preconditions.checkPositionIndexes(Preconditions.java:1430)
    at com.google.common.collect.ImmutableList.subList(ImmutableList.java:450)
    at com.google.devtools.build.skyframe.InvalidatingNodeVisitor$DeletingNodeVisitor.lambda$runInternal$0(InvalidatingNodeVisitor.java:280)
    at com.google.devtools.build.lib.concurrent.AbstractQueueVisitor$WrappedRunnable.run(AbstractQueueVisitor.java:365)
    at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1426)
    at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
    at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
    at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
    at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
    at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
```

**Comment by iancha1992** [comment]
@bazel-io fork 7.5.0

**Comment by iancha1992** [comment]
A fix for this issue has been included in [Bazel 7.5.0 RC2](https://github.com/bazelbuild/bazel/releases/tag/7.5.0rc2). Please test out the release candidate and report any issues as soon as possible.
If you're using Bazelisk, you can point to the latest RC by setting USE_BAZEL_VERSION=7.5.0rc2. Thanks!
